### ISSUE TITLE
CloudFlare R2 storage? (S3 compatible, zero egress fees)

### ISSUE THREAD

**Issue by turian** [skypilot-org/skypilot#1639]
Would it be possible to add support for CloudFlare R2 storage?

It is S3 compatible and there are zero egress fees.

**Comment by concretevitamin** [comment]
@turian thanks. R2 is definitely interesting to us. Are you looking to have a SkyPilot job read from an R2 bucket (training data?)? What's the current workaround?

**Comment by romilbhardwaj** [comment]
Yes, it seems feasible. At a high level, we need two key components to implement a `R2Store` [class](https://github.com/skypilot-org/skypilot/blob/e535f29546fe905eb8d651a08fc436fbfce68ff1/sky/data/storage.py#L108):

1. A method to upload files - seems like R2 can use [aws cli](https://developers.cloudflare.com/r2/examples/aws-cli/), which we already use in our `S3Store` implementation.
2. A method to mount directories - since R2 exposes a S3 api, we might be able to use [goofys](https://github.com/skypilot-org/skypilot/blob/e535f29546fe905eb8d651a08fc436fbfce68ff1/sky/data/storage.py#L1004) like we do for `S3Store`.

We would greatly appreciate any help with this - would you be interested in contributing R2 support?

**Comment by turian** [comment]
@concretevitamin Yes, read from R2 during training and possibly write checkpoints etc to R2.

@romilbhardwaj Unfortunately I can't help on this currently

**Comment by landscapepainter** [comment]
@turian We just completed our [first integration of r2](https://github.com/skypilot-org/skypilot/commit/01e98e826609d513bfd4b49c8f79c9f3f0f9501d). Try out and please let us know what you think! The instruction to set up r2 for skypilot is [here](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#cloudflare-r2) as well.

**Comment by turian** [comment]
@landscapepainter Thank you. At this current point in my experimental cycle I cannot test immediately. I still think this is quite a valuable feature that should be first class for many users.
